[BUG] v1.2.1 breaks usage of baker.make(related_model__field=value)

Specifying field values of related model instances within the `baker.make()` signature (e.g. `baker.make(my_related_model__field=value)` is not possible anymore with v1.2.1. This still works in v1.2.0. 

## Expected behavior

The model instance AND its required related model instances should be created, populated with random defaults except the values we explicitly passed in the call.

## Actual behavior

`ValueError: Cannot assign "<fk-id>": "<model>.<fk-field>" must be a "<fk-model>" instance.`

## Reproduction Steps

Try creating a model instance with `baker.make()` using related field assignments **for an FK field that has a default value/callable returning an ID** as seen in the following example. Any related models will do, these are just for illustration:

```python
# models.py

def get_default_category():
    """ 
    Returning an ID here will make model_bakery fail, returning an instance is fine, but see:
    https://docs.djangoproject.com/en/3.1/ref/models/fields/#django.db.models.Field.default 
    """
    return 1


class SpotCategory(models.Model):
    name = models.CharField(max_length=128)


class TVSpot(models.Model):
    category = models.ForeignKey(SpotCategory, on_delete=models.CASCADE, default=get_default_category)


# tests.py

class TestBugWithModelBakeryVersion_1_2_1:

    def test_creating_related_models_separately_works(self, db):
        """This seems to always work, no matter the version of model_bakery."""

        category = baker.make("SpotCategory", name="Special")
        baker.make("TVSpot", category=category)

    def test_specifying_related_model_fields_within_constructor_fails_with_v1_2_1(
        self, db
    ):
        """This worked up until 1.2.0, but fails with 1.2.1 as follows:

        ValueError: Cannot assign "1990": "TVSpot.category" must be a "SpotCategory" instance.

        """
        baker.make("TVSpot", category__name="Special")

```

Run tests with different model_bakery versions like:

```bash

$ pip install model_bakery==1.2.0 && pytest -k TestBugWithModelBakeryVersion_1_2_1  # should pass
$ pip install model_bakery==1.2.1 && pytest -k TestBugWithModelBakeryVersion_1_2_1  # should fail

```

### Versions

OS: Ubuntu 20.04
Database: MySQL 5.7.32
Python: 3.8
Django: 2.2.17
pytest: 6.1.2
Model Bakery: 1.2.1

Maybe you do have an idea which of the changes in 1.2.1 is related to this. Let me know what I can do to help fix this, thanks.

Full pytest output:

```
===================================================================================================== FAILURES ======================================================================================================
__________________________________________________________ test_create_tvspot_with_specific_category __________________________________________________________

    def test_create_tvspot_with_specific_category(db):
>       assert baker.make(
            "TVSpot", category__name="Name"
        ).some_flag

msrcms/api/tests/test_misc.py:2274: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../.virtualenvs/venv/lib/python3.8/site-packages/model_bakery/baker.py:80: in make
    return baker.make(
../../.virtualenvs/venv/lib/python3.8/site-packages/model_bakery/baker.py:302: in make
    return self._make(**params)
../../.virtualenvs/venv/lib/python3.8/site-packages/model_bakery/baker.py:357: in _make
    instance = self.instance(
../../.virtualenvs/venv/lib/python3.8/site-packages/model_bakery/baker.py:388: in instance
    instance = self.model(**attrs)  # type: Model
msrcms/utils/__init__.py:479: in new_init
    new_init._original_init(self, *args, **kwargs)
../../.virtualenvs/cms/lib/python3.8/site-packages/django/db/models/base.py:483: in __init__
    _setattr(self, field.name, rel_obj)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <django.db.models.fields.related_descriptors.ForwardManyToOneDescriptor object at 0x7f73ec861070>
instance = <TVSpot: wAtkPGavzuyoryTgNxCgfaCiJMYlnRAESEHodYenUmcxldioNl | VgluBXBRPnczSqzDcuenBJRnokEUIDBvYzhsUkcrflgMjKfLdiszTOaL...jyQgyBCwMzIYtuSRTCSopKqyDHrTqtEsFfWrRIruxvbdwfvnuMFXoCvfNwmlqxenAkNlswTkrgLFwdUTuUInBGcmvgBptoKczSidMeceFvdFTNXpdAiGBO>
value = 2586

    def __set__(self, instance, value):
        """
        Set the related instance through the forward relation.
    
        With the example above, when setting ``child.parent = parent``:
    
        - ``self`` is the descriptor managing the ``parent`` attribute
        - ``instance`` is the ``child`` instance
        - ``value`` is the ``parent`` instance on the right of the equal sign
        """
        # An object must be an instance of the related class.
        if value is not None and not isinstance(value, self.field.remote_field.model._meta.concrete_model):
>           raise ValueError(
                'Cannot assign "%r": "%s.%s" must be a "%s" instance.' % (
                    value,
                    instance._meta.object_name,
                    self.field.name,
                    self.field.remote_field.model._meta.object_name,
                )
E               ValueError: Cannot assign "2586": "TVSpot.category" must be a "SpotCategory" instance.

```

I found issue that might be related:
If field has explicit `default`, `_fill_optional=True` now fills it with that default value:
```
class Issue(models.Model):
    comment = models.TextField(default='', blank=True)
```
then in test
```
        issue = baker.make('Issue', _fill_optional=True)
>       assert issue.comment
E       AssertionError: assert ''
```
It wasn't the case before 1.2.1.
If you remove `default=''`, the test passes.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -18,7 +18,8 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 ### Changed
 
 - Type hinting fixed for Recipe "_model" parameter  [PR #124](https://github.com/model-bakers/model_bakery/pull/124)
-- Modify `setup.py` to not import the whole module for package data, but get it from `__about__.py`  [PR #142](https://github.com/model-bakers/model_bakery/pull/142)
+- Fixed a bug (introduced in 1.2.1) that was breaking creation of model instances with related model fields [PR #164](https://github.com/model-bakers/model_bakery/pull/164)
+- [dev] Modify `setup.py` to not import the whole module for package data, but get it from `__about__.py`  [PR #142](https://github.com/model-bakers/model_bakery/pull/142)
 - [dev] Add Dependabot config file [PR #146](https://github.com/model-bakers/model_bakery/pull/146)
 
 ### Removed

--- a/model_bakery/baker.py
+++ b/model_bakery/baker.py
@@ -349,23 +349,7 @@ class Baker(object):
                     self.m2m_dict[field.name] = self.m2m_value(field)
                 else:
                     self.m2m_dict[field.name] = self.model_attrs.pop(field.name)
-
             elif field.name not in self.model_attrs:
-                # TODO: this might be a code going into a wrong direction, drop it if
-                #  things in `model_bakery.baker.Baker.generate_value` will work out
-                # if (
-                #     isinstance(field, ForeignKey)
-                #     and "{0}_id".format(field.name) not in self.model_attrs
-                # ):
-                #     value = self.generate_value(field, commit_related)
-                #     if isinstance(value, field.related_model):
-                #         field_name = field.name
-                #     else:
-                #         field_name = "{0}_id".format(field.name)
-                #
-                #     self.model_attrs[field_name] = value
-                #
-                # elif (
                 if (
                     not isinstance(field, ForeignKey)
                     or "{0}_id".format(field.name) not in self.model_attrs
@@ -547,10 +531,11 @@ class Baker(object):
         """Call the associated generator with a field passing all required args.
 
         Generator Resolution Precedence Order:
-        -- attr_mapping - mapping per attribute name
-        -- choices -- mapping from avaiable field choices
-        -- type_mapping - mapping from user defined type associated generators
-        -- default_mapping - mapping from pre-defined type associated
+        -- `field.default` - model field default value, unless explicitly overwritten during baking
+        -- `attr_mapping` - mapping per attribute name
+        -- `choices` -- mapping from available field choices
+        -- `type_mapping` - mapping from user defined type associated generators
+        -- `default_mapping` - mapping from pre-defined type associated
            generators
 
         `attr_mapping` and `type_mapping` can be defined easily overwriting the
@@ -559,7 +544,7 @@ class Baker(object):
         is_content_type_fk = isinstance(field, ForeignKey) and issubclass(
             self._remote_field(field).model, contenttypes.models.ContentType
         )
-        # when a default provided, we only use it unless the field is listed in `self.rel_fields`
+        # we only use default unless the field is overwritten in `self.rel_fields`
         if field.has_default() and field.name not in self.rel_fields:
             if callable(field.default):
                 return field.default()

--- a/model_bakery/baker.py
+++ b/model_bakery/baker.py
@@ -349,8 +349,21 @@ class Baker(object):
                     self.m2m_dict[field.name] = self.m2m_value(field)
                 else:
                     self.m2m_dict[field.name] = self.model_attrs.pop(field.name)
+
             elif field.name not in self.model_attrs:
                 if (
+                    isinstance(field, ForeignKey)
+                    and "{0}_id".format(field.name) not in self.model_attrs
+                ):
+                    value = self.generate_value(field, commit_related)
+                    if isinstance(value, field.related_model):
+                        field_name = field.name
+                    else:
+                        field_name = "{0}_id".format(field.name)
+
+                    self.model_attrs[field_name] = value
+
+                elif (
                     not isinstance(field, ForeignKey)
                     or "{0}_id".format(field.name) not in self.model_attrs
                 ):

--- a/model_bakery/baker.py
+++ b/model_bakery/baker.py
@@ -351,19 +351,22 @@ class Baker(object):
                     self.m2m_dict[field.name] = self.model_attrs.pop(field.name)
 
             elif field.name not in self.model_attrs:
+                # TODO: this might be a code going into a wrong direction, drop it if
+                #  things in `model_bakery.baker.Baker.generate_value` will work out
+                # if (
+                #     isinstance(field, ForeignKey)
+                #     and "{0}_id".format(field.name) not in self.model_attrs
+                # ):
+                #     value = self.generate_value(field, commit_related)
+                #     if isinstance(value, field.related_model):
+                #         field_name = field.name
+                #     else:
+                #         field_name = "{0}_id".format(field.name)
+                #
+                #     self.model_attrs[field_name] = value
+                #
+                # elif (
                 if (
-                    isinstance(field, ForeignKey)
-                    and "{0}_id".format(field.name) not in self.model_attrs
-                ):
-                    value = self.generate_value(field, commit_related)
-                    if isinstance(value, field.related_model):
-                        field_name = field.name
-                    else:
-                        field_name = "{0}_id".format(field.name)
-
-                    self.model_attrs[field_name] = value
-
-                elif (
                     not isinstance(field, ForeignKey)
                     or "{0}_id".format(field.name) not in self.model_attrs
                 ):
@@ -556,8 +559,8 @@ class Baker(object):
         is_content_type_fk = isinstance(field, ForeignKey) and issubclass(
             self._remote_field(field).model, contenttypes.models.ContentType
         )
-
-        if field.has_default():
+        # when a default provided, we only use it unless the field is listed in `self.rel_fields`
+        if field.has_default() and field.name not in self.rel_fields:
             if callable(field.default):
                 return field.default()
             return field.default

--- a/tests/generic/models.py
+++ b/tests/generic/models.py
@@ -271,6 +271,15 @@ class DummyForeignKeyWithDefaultIdModel(models.Model):
     )
 
 
+class DummyOptionalForeignKey(models.Model):
+    named_thing = models.ForeignKey(
+        "NamedThing",
+        null=True,
+        default=None,
+        on_delete=models.SET_NULL,
+    )
+
+
 class DummyNullFieldsModel(models.Model):
     null_foreign_key = models.ForeignKey(
         "DummyBlankFieldsModel", null=True, on_delete=models.CASCADE

--- a/tests/generic/models.py
+++ b/tests/generic/models.py
@@ -254,6 +254,23 @@ class DummyGenericRelationModel(models.Model):
     relation = GenericRelation(DummyGenericForeignKeyModel)
 
 
+class NamedThing(models.Model):
+    name = models.CharField(max_length=64)
+
+
+def get_default_namedthing_id():
+    instance, _ = NamedThing.objects.get_or_create(name="Default")
+    return instance.id
+
+
+class DummyForeignKeyWithDefaultIdModel(models.Model):
+    named_thing = models.ForeignKey(
+        "NamedThing",
+        default=get_default_namedthing_id,
+        on_delete=models.SET_DEFAULT,
+    )
+
+
 class DummyNullFieldsModel(models.Model):
     null_foreign_key = models.ForeignKey(
         "DummyBlankFieldsModel", null=True, on_delete=models.CASCADE

--- a/tests/generic/models.py
+++ b/tests/generic/models.py
@@ -181,6 +181,10 @@ class LonelyPerson(models.Model):
     only_friend = models.OneToOneField(Person, on_delete=models.CASCADE)
 
 
+class Cake(models.Model):
+    name = models.CharField(max_length=64)
+
+
 class RelatedNamesModel(models.Model):
     name = models.CharField(max_length=256)
     one_to_one = models.OneToOneField(
@@ -188,6 +192,30 @@ class RelatedNamesModel(models.Model):
     )
     foreign_key = models.ForeignKey(
         Person, related_name="fk_related", on_delete=models.CASCADE
+    )
+
+
+def get_default_cake_id():
+    instance, _ = Cake.objects.get_or_create(name="Muffin")
+    return instance.id
+
+
+class RelatedNamesWithDefaultsModel(models.Model):
+    name = models.CharField(max_length=256, default="Bravo")
+    cake = models.ForeignKey(
+        Cake,
+        on_delete=models.CASCADE,
+        default=get_default_cake_id,
+    )
+
+
+class RelatedNamesWithEmptyDefaultsModel(models.Model):
+    name = models.CharField(max_length=256, default="Bravo")
+    cake = models.ForeignKey(
+        Cake,
+        on_delete=models.CASCADE,
+        null=True,
+        default=None,
     )
 
 
@@ -252,32 +280,6 @@ class DummyGenericForeignKeyModel(models.Model):
 
 class DummyGenericRelationModel(models.Model):
     relation = GenericRelation(DummyGenericForeignKeyModel)
-
-
-class NamedThing(models.Model):
-    name = models.CharField(max_length=64)
-
-
-def get_default_namedthing_id():
-    instance, _ = NamedThing.objects.get_or_create(name="Default")
-    return instance.id
-
-
-class DummyForeignKeyWithDefaultIdModel(models.Model):
-    named_thing = models.ForeignKey(
-        "NamedThing",
-        default=get_default_namedthing_id,
-        on_delete=models.SET_DEFAULT,
-    )
-
-
-class DummyOptionalForeignKey(models.Model):
-    named_thing = models.ForeignKey(
-        "NamedThing",
-        null=True,
-        default=None,
-        on_delete=models.SET_NULL,
-    )
 
 
 class DummyNullFieldsModel(models.Model):

--- a/tests/test_filling_fields.py
+++ b/tests/test_filling_fields.py
@@ -279,24 +279,32 @@ class TestFillingGenericForeignKeyField:
 @pytest.mark.django_db
 class TestFillingForeignKeyFieldWithDefaultFunctionReturningId:
     def test_filling_foreignkey_with_default_id(self):
-        dummy = baker.make(models.DummyForeignKeyWithDefaultIdModel)
-        assert dummy.named_thing.id == models.get_default_namedthing_id()
-        assert dummy.named_thing.name == "Default"
+        dummy = baker.make(models.RelatedNamesWithDefaultsModel)
+        assert dummy.cake.__class__.objects.count() == 1
+        assert dummy.cake.id == models.get_default_cake_id()
+        assert dummy.cake.name == "Muffin"
 
     def test_filling_foreignkey_with_default_id_with_custom_arguments(self):
         dummy = baker.make(
-            models.DummyForeignKeyWithDefaultIdModel, named_thing__name="Not default"
+            models.RelatedNamesWithDefaultsModel, cake__name="Baumkuchen"
         )
-        assert dummy.named_thing.__class__.objects.count() == 1
-        assert dummy.named_thing.id == dummy.named_thing.__class__.objects.get().id
-        assert dummy.named_thing.name == "Not default"
+        assert dummy.cake.__class__.objects.count() == 1
+        assert dummy.cake.id == dummy.cake.__class__.objects.get().id
+        assert dummy.cake.name == "Baumkuchen"
 
 
 @pytest.mark.django_db
 class TestFillingOptionalForeignKeyField:
+    def test_not_filling_optional_foreignkey(self):
+        dummy = baker.make(models.RelatedNamesWithEmptyDefaultsModel)
+        assert dummy.cake is None
+
     def test_filling_optional_foreignkey_implicitly(self):
-        dummy = baker.make(models.DummyOptionalForeignKey, named_thing__name="Optional")
-        assert dummy.named_thing.name == "Optional"
+        dummy = baker.make(
+            models.RelatedNamesWithEmptyDefaultsModel, cake__name="Carrot cake"
+        )
+        assert dummy.cake.__class__.objects.count() == 1
+        assert dummy.cake.name == "Carrot cake"
 
 
 @pytest.mark.django_db

--- a/tests/test_filling_fields.py
+++ b/tests/test_filling_fields.py
@@ -277,6 +277,15 @@ class TestFillingGenericForeignKeyField:
 
 
 @pytest.mark.django_db
+class TestFillingForeignKeyFieldWithDefaultFunctionReturningId:
+    def test_filling_foreignkey_with_default_id(self):
+        dummy = baker.make(
+            models.DummyForeignKeyWithDefaultIdModel, named_thing__name="Default"
+        )
+        assert dummy.named_thing.id == models.get_default_namedthing_id()
+
+
+@pytest.mark.django_db
 class TestsFillingFileField:
     def test_filling_file_field(self):
         dummy = baker.make(models.DummyFileFieldModel, _create_files=True)

--- a/tests/test_filling_fields.py
+++ b/tests/test_filling_fields.py
@@ -279,10 +279,24 @@ class TestFillingGenericForeignKeyField:
 @pytest.mark.django_db
 class TestFillingForeignKeyFieldWithDefaultFunctionReturningId:
     def test_filling_foreignkey_with_default_id(self):
-        dummy = baker.make(
-            models.DummyForeignKeyWithDefaultIdModel, named_thing__name="Default"
-        )
+        dummy = baker.make(models.DummyForeignKeyWithDefaultIdModel)
         assert dummy.named_thing.id == models.get_default_namedthing_id()
+        assert dummy.named_thing.name == "Default"
+
+    def test_filling_foreignkey_with_default_id_with_custom_arguments(self):
+        dummy = baker.make(
+            models.DummyForeignKeyWithDefaultIdModel, named_thing__name="Not default"
+        )
+        assert dummy.named_thing.__class__.objects.count() == 1
+        assert dummy.named_thing.id == dummy.named_thing.__class__.objects.get().id
+        assert dummy.named_thing.name == "Not default"
+
+
+@pytest.mark.django_db
+class TestFillingOptionalForeignKeyField:
+    def test_filling_optional_foreignkey_implicitly(self):
+        dummy = baker.make(models.DummyOptionalForeignKey, named_thing__name="Optional")
+        assert dummy.named_thing.name == "Optional"
 
 
 @pytest.mark.django_db

